### PR TITLE
feat(engine): use execute_with_state_hook in ExecutorMetrics

### PR DESCRIPTION
**Pull request by fgimenez**
Towards #12053

Preparing ground to integrate state root task with executor, this PR includes changes to allow the introduction of a state hook during block execution in engine tree, so that the `StateRootTask` can receive state updates after each transaction execution.

The changes in `ExecutorMetrics` are mainly about getting execution metrics using `execute_with_state_hook` instead of `execute_with_state_closure` as before. 

Added tests to cover `ExecutorMetrics` changes, metrics still recorded and state hook being called.

### CODE REVIEW

File: crates/evm/src/metrics.rs [hunk 1]
```diff
@@ -2,14 +2,41 @@
...
+use crate::{execute::Executor, system_calls::OnStateHook};
 use metrics::{Counter, Gauge, Histogram};
 use reth_execution_types::{BlockExecutionInput, BlockExecutionOutput};
 use reth_metrics::Metrics;
 use reth_primitives::BlockWithSenders;
+use revm_primitives::ResultAndState;
+use std::time::Instant;
+
+/// Wrapper struct that combines metrics and state hook
+struct StateHookWrapper {
```
Comment (mattsse) [inline crates/evm/src/metrics.rs:14]: can we rename this to smth like `MeteredStateHook`?

**Review by mattsse — APPROVED** [review]
lgtm, naming suggestion

File: crates/evm/src/metrics.rs (hunk 1 above)
Comment (fgimenez) [reply crates/evm/src/metrics.rs]: good call, done here https://github.com/paradigmxyz/reth/pull/12316/commits/324d4ff5d5568b83ac285c17a4ef49f463cdcd57

**Review by rkrasiuk — APPROVED** [review]
lgtm

**Pull request merged**